### PR TITLE
OrderedCollection iterator and Vec conversion

### PR DESCRIPTION
**Pull request by bazhenov**
PR regarding https://github.com/jonhoo/ordsearch/pull/27#issuecomment-1879672446.

This implementation provides:

- `OrderedCollection::to_vec()`
- `OrderedCollection::iter()`
- `From` implementation for converting `OrderedCollection` to `Vec`

The original plan of implementing `into_iter()` doesn't work. We cannot implement `into_iter()` for `OC` because we cannot move out of `self.items` because `OC` implements custom `Drop`. So we are forced to create a copy of items anyway. In this case If the user wants data in a `Vec`, all items will be copied twice when using `into_iter()`. In this case IMO it is more appropriate to implement `to_vec()`.

### CODE REVIEW

**Comment by jonhoo** [comment]
Hmm, I think `into_iter` should be feasible by doing a mem::take to get the Vec out of `self`. And then add a condition on `!is_empty` in Drop.

**Comment by bazhenov** [comment]
Indeed, thank you! No need to change `drop()` implementation. `truncate()` is totally fine with indicies larger than size of the vector.

File: src/lib.rs [hunk 1]
```diff
@@ -395,6 +395,83 @@ impl<T: Ord> OrderedCollection<T> {
...
+    /// let mut values: Vec<_> = coll.iter().copied().collect();
+    /// values.sort();
+    /// assert_eq!(values, expected);
+    /// ```
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
+        Iter { coll: self, idx: 0 }
+    }
+}
+
+impl<T: Clone> OrderedCollection<T> {
```
Comment (jonhoo) [inline src/lib.rs]: I don't think we need the `Clone` bound any more?

File: src/lib.rs [hunk 2]
```diff
@@ -395,6 +395,83 @@ impl<T: Ord> OrderedCollection<T> {
...
+    ///
+    /// ```
+    /// # use ordsearch::OrderedCollection;
+    /// let x = vec![1, 2, 3, 4, 5];
+    /// let coll = OrderedCollection::from(x.clone());
+    /// let mut values: Vec<_> = coll.into_vec();
+    /// values.sort();
+    /// assert_eq!(values, x);
+    /// ```
+    pub fn into_vec(mut self) -> Vec<T> {
```
Comment (jonhoo) [inline src/lib.rs]: I'm tempted to not have this method and _just_ have the `From`, but don't feel too strongly — what do you think?

File: src/lib.rs [hunk 3]
```diff
@@ -395,6 +395,83 @@ impl<T: Ord> OrderedCollection<T> {
...
+        assert!(!self.items.is_empty());
+
+        let mut items = mem::take(&mut self.items);
+        items.swap_remove(0);
+        // SAFETY: 0-th element already removed, so all initialized
+        unsafe { mem::transmute(items) }
+    }
+}
+
+impl<T: Clone> IntoIterator for OrderedCollection<T> {
```
Comment (jonhoo) [inline src/lib.rs]: Ditto here: no need for `Clone`.

File: src/lib.rs [hunk 4]
```diff
@@ -395,6 +395,83 @@ impl<T: Ord> OrderedCollection<T> {
...
+    ///
+    /// ```
+    /// # use ordsearch::OrderedCollection;
+    /// let expected = vec![1, 2, 3, 4, 5];
+    /// let coll = OrderedCollection::from(expected.clone());
+    /// let mut values: Vec<_> = coll.iter().copied().collect();
+    /// values.sort();
+    /// assert_eq!(values, expected);
+    /// ```
+    pub fn iter(&self) -> impl Iterator<Item = &T> {
```
Comment (jonhoo) [inline src/lib.rs]: Let's also `impl IntoIterator for &OrderedCollection`

File: src/lib.rs [hunk 5]
```diff
@@ -395,6 +395,83 @@ impl<T: Ord> OrderedCollection<T> {
...
+    idx: usize,
+}
+
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.idx += 1;
+        if self.idx < self.coll.items.len() {
+            let value = &self.coll.items[self.idx];
```
Comment (jonhoo) [inline src/lib.rs:450]: This can use `.get_unchecked` as well I think for the same safety reasons as below.

File: src/lib.rs [hunk 6]
```diff
@@ -395,6 +395,83 @@ impl<T: Ord> OrderedCollection<T> {
...
+            // SAFETY: i > 0, so only initialized items are accessed
+            // SAFETY: i < self.coll.items.len() so no out-of-bounds access
+            Some(unsafe { value.assume_init_ref() })
+        } else {
+            None
+        }
+    }
+}
+
+impl<T: Clone> From<OrderedCollection<T>> for Vec<T> {
```
Comment (jonhoo) [inline src/lib.rs]: Ditto re `Clone`.

File: src/lib.rs [hunk 7]
```diff
@@ -517,6 +594,16 @@ mod tests {
         assert_eq!(x.find_gte(257), None);
     }
 
+    #[test]
+    fn check_into_iter() {
```
Comment (jonhoo) [inline src/lib.rs:604]: Can you also add a test for `.iter`?

**Review by jonhoo — CHANGES_REQUESTED** [review]
:tada:

File: src/lib.rs (hunk 6 above)
Comment (bazhenov) [reply src/lib.rs]: ✅

File: src/lib.rs (hunk 4 above)
Comment (bazhenov) [reply src/lib.rs]: ✅

File: src/lib.rs (hunk 3 above)
Comment (bazhenov) [reply src/lib.rs]: Good catch! ✅

File: src/lib.rs (hunk 1 above)
Comment (bazhenov) [reply src/lib.rs]: ✅

File: src/lib.rs (hunk 2 above)
Comment (bazhenov) [reply src/lib.rs]: The same. I was focused on how `slice` is implemented – it has both `From` and `into_vec()` implementations, but I'm not sure if there is any specific reason behind it.

Let's remove it.

File: src/lib.rs (hunk 5 above)
Comment (bazhenov) [reply src/lib.rs:450]: Let me do some benchmarking to test if it helps in this particular case and if we may get away without adding an unsafe code.

I have a dubious suspicion that because we already test `idx` for iterator termination criteria, the compiler can eliminate duplicating index checks.

File: src/lib.rs (hunk 7 above)
Comment (bazhenov) [reply src/lib.rs:604]: Just to make sure. There is already a doc-test on `OC::iter()`. Do you want to have additional explicit test?

Comment (jonhoo) [reply src/lib.rs:604]: Yeah, always handy to have an actual unit test in addition to the doc tests. For example, doc tests currently (unless something has changed recently) don't get counted towards coverage. Also, doc tests tend to focus on "how does this API work" rather than "let's push on some of the corner cases". For instance, would be good to test that iteration also works correctly on an empty collection.

Comment (bazhenov) [reply src/lib.rs:604]: ✅

File: src/lib.rs [hunk 8]
```diff
@@ -517,6 +600,39 @@ mod tests {
...
+            .copied()
+            .collect::<Vec<_>>();
+        values.sort();
+        assert_eq!(values, expected);
+    }
+
+    #[test]
+    fn check_iter_empty() {
+        let values = OrderedCollection::<u32>::from(vec![]);
+        assert!(values.iter().next().is_none());
```
Comment (jonhoo) [inline src/lib.rs]:
```suggestion
        assert_eq!(values.iter().next(), None);
```
Gives slightly more helpful errors. Ditto in the other test.

Comment (bazhenov) [reply src/lib.rs]: done

File: src/lib.rs (hunk 5 above)
Comment (bazhenov) [reply src/lib.rs:450]: I tried following implementation of a `next()`:

```rust
    fn next(&mut self) -> Option<Self::Item> {
        self.idx += 1;
        if self.idx < self.coll.items.len() {
            let value = unsafe { self.coll.items.get_unchecked(self.idx) };
            // SAFETY: i > 0, so only initialized items are accessed
            // SAFETY: i < self.coll.items.len() so no out-of-bounds access
            Some(unsafe { value.assume_init_ref() })
        } else {
            None
        }
    }
```

Criterion doesn't see any significant difference:

```
Iteration/Iteration     time:   [329.94 µs 330.98 µs 332.13 µs]
                        change: [+0.0272% +0.7601% +1.5008%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe
```

If you put `self.coll.items.len()` into a blackbox preventing compiler optimizations, there is around 3% improvement when using `get_unchecked()` (comparing to the default variant with blackbox) which is consistent with my theory:

```
Iteration/Iteration     time:   [428.03 µs 428.42 µs 428.83 µs]
                        change: [-4.0191% -3.7239% -3.4410%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
```

I'm advocating for not using unsafe in this scenario.

P.S. let me know if you want me to add benchmarks to the PR.

Comment (bazhenov) [reply src/lib.rs:450]: btw, quite probably there is also no need for `get_unchecked()` in a `find_gte()` for the very same reason. But it is out of scope of this PR.

Comment (jonhoo) [reply src/lib.rs:450]: Nice, thanks for checking!

**Review by jonhoo — APPROVED** [review]

**Comment by jonhoo** [comment]
Thanks! Will publish probably later today 👍

**Comment by jonhoo** [comment]
Released in 0.2.7 :tada:
